ethers-core/src/utils/geth.rs

```diff
@@ -93,10 +154,54 @@ impl Geth {
         self
     }
 
+    /// Sets the port which will be used for incoming p2p connections.
+    ///
+    /// This will put the geth instance into non-dev mode, discarding any previously set dev-mode
+    /// options.
+    #[must_use]
+    pub fn p2p_port(mut self, port: u16) -> Self {
+        match self.mode {
+            GethMode::Dev(_) => {
+                self.mode = GethMode::NonDev(PrivateNetOptions {
+                    p2p_port: Some(port),
+                    ..Default::default()
+                })
+            }
+            GethMode::NonDev(ref mut opts) => opts.p2p_port = Some(port),
+        }
+        self
+    }
+
     /// Sets the block-time which will be used when the `geth-cli` instance is launched.
+    ///
+    /// This will put the geth instance in `dev` mode, discarding any previously set options that
+    /// cannot be used in dev mode.
     #[must_use]
     pub fn block_time<T: Into<u64>>(mut self, block_time: T) -> Self {
-        self.block_time = Some(block_time.into());
+        self.mode = GethMode::Dev(DevOptions { block_time: Some(block_time.into()) });
+        self
+    }
+
+    /// Sets the chain id for the geth instance.
+    #[must_use]
+    pub fn chain_id<T: Into<u64>>(mut self, chain_id: T) -> Self {
+        self.chain_id = Some(chain_id.into());
+        self
+    }
+
+    /// Disable discovery for the geth instance.
+    ///
+    /// This will put the geth instance into non-dev mode, discarding any previously set dev-mode
+    /// options.
+    #[must_use]
+    pub fn disable_discovery(mut self) -> Self {
```

Comment:
nice. we should consider thinking of how can we use these options on geth to cause partitions intentionally and see how our failure recovery works